### PR TITLE
FPLA-803: Amend-Behaviour-Of-C6-C6A-Tick-Boxes

### PR DESCRIPTION
**Pull request by Jamesnelson117**
**Before creating a pull request make sure that:**

- [x] commit messages are meaningful and follow good commit message guidelines
- [x] README and other documentation has been updated / added (if needed)
- [x] tests have been updated / new tests has been added (if needed)

### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/FPLA-803

### Change description ###
Fixes bug in which notice of proceedings where being removed if pre-existing on caseData but not selected to be re-generated. 

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsController.java [hunk 1]
```diff
@@ -104,17 +104,45 @@ public AboutToStartOrSubmitCallbackResponse handleAboutToSubmitEvent(
...
+
+        if (callbackRequest.getCaseDetailsBefore() != null
+            && callbackRequest.getCaseDetailsBefore().getData().get("noticeOfProceedingsBundle") != null) {
+            CaseData caseDataBefore = mapper.convertValue(callbackRequest.getCaseDetailsBefore().getData(),
+                CaseData.class);
+
+            noticeOfProceedingCaseData.addAll(getRemovedDocumentBundles(caseDataBefore, templateTypes));
+        }
+
+        caseDetails.getData().put("noticeOfProceedingsBundle", noticeOfProceedingCaseData);
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsController.java]: Couldn't that be extracted to the service? The integration tests would be easier for you and you could focus more on unit testing.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsController.java [hunk 2]
```diff
@@ -104,17 +104,45 @@ public AboutToStartOrSubmitCallbackResponse handleAboutToSubmitEvent(
...
+        List<String> templateTypeTitles = templateTypes.stream().map(DocmosisTemplates::getDocumentTitle)
+            .collect(Collectors.toList());
+
+        ImmutableList.Builder<Element> removedDocumentBundles = ImmutableList.builder();
+
+        caseData.getNoticeOfProceedingsBundle().forEach(element -> {
+            String filename = element.getValue().getDocument().getFilename();
+            filename = filename.substring(0, filename.lastIndexOf("."));
+
+            if (templateTypeTitles.stream().parallel().noneMatch(filename::equals)) {
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsController.java]: Maybe just if (!templateTypeTitles.contains(filename))?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsController.java [hunk 3]
```diff
@@ -104,17 +104,45 @@ public AboutToStartOrSubmitCallbackResponse handleAboutToSubmitEvent(
...
 
+    private List<Element> getRemovedDocumentBundles(CaseData caseData, List<DocmosisTemplates> templateTypes) {
+        List<String> templateTypeTitles = templateTypes.stream().map(DocmosisTemplates::getDocumentTitle)
+            .collect(Collectors.toList());
+
+        ImmutableList.Builder<Element> removedDocumentBundles = ImmutableList.builder();
+
+        caseData.getNoticeOfProceedingsBundle().forEach(element -> {
+            String filename = element.getValue().getDocument().getFilename();
+            filename = filename.substring(0, filename.lastIndexOf("."));
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsController.java]: Maybe there is something from FileNameUtils? Like FileNameUtils.getName?

File: ccd-definition/AuthorisationCaseEvent.json [hunk 4]
```diff
@@ -327,5 +327,12 @@
     "CaseEventID": "addStatementOfService",
     "UserRole": "caseworker-publiclaw-solicitor",
     "CRUD": "CRU"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseEventID": "createNoticeOfProceedings",
+    "UserRole": "caseworker-publiclaw-courtadmin",
+    "CRUD": "CRU"
```
[the remark this answers is not in the document]
Comment (Jamesnelson117) [reply ccd-definition/AuthorisationCaseEvent.json]: No, will need removed again before merge.

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsControllerAboutToSubmitTest.java [hunk 5]
```diff
@@ -97,13 +101,103 @@ void shouldGenerateC6NoticeOfProceedingsDocument() throws Exception {
...
+
+        Map<String, String> templateData = createTemplatePlaceholders();
+
+        given(caseDataExtractionService.getNoticeOfProceedingTemplateData(caseData))
+            .willReturn(templateData);
+        given(docmosisDocumentGeneratorService.generateDocmosisDocument(templateData, C6))
+            .willReturn(docmosisDocument);
+        given(uploadDocumentService.uploadPDF(USER_ID, AUTH_TOKEN, PDF, C6_DOCUMENT_TITLE + ".pdf"))
+            .willReturn(document);
+
```
[the remark this answers is not in the document]
Comment (Jamesnelson117) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsControllerAboutToSubmitTest.java]: Yes, let's move getRemovedDocumentBundle() to it's own service and move a lot of these tests to service tests.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsController.java (hunk 1 above)
Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsController.java]: Extracted to NoticeOfProceedingService

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsController.java (hunk 3 above)
Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsController.java]: fileNameUtils.getBaseName(String) is what we needed :)

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsController.java [hunk 6]
```diff
@@ -105,7 +109,18 @@ public AboutToStartOrSubmitCallbackResponse handleAboutToSubmitEvent(
         List<Document> uploadedDocuments = generateAndUploadDocuments(userId, authorization, templateData,
             templateTypes);
 
-        caseDetails.getData().put("noticeOfProceedingsBundle", createNoticeOfProceedingsCaseData(uploadedDocuments));
+        List<Element> noticeOfProceedingCaseData = createNoticeOfProceedingsCaseData(uploadedDocuments);
```
[the remark this answers is not in the document]
Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsController.java]: Look at latest, this has changed.

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsControllerAboutToSubmitTest.java [hunk 7]
```diff
@@ -56,28 +57,29 @@
...
+            .bytes(PDF)
             .documentTitle(C6_DOCUMENT_TITLE)
             .build();
 
-        CallbackRequest caseDetails = mapper.readValue(readBytes("fixtures/C6CaseData.json"),
-            CallbackRequest.class);
+        CallbackRequest callbackRequest = CallbackRequest.builder()
+            .caseDetails(callbackRequest().getCaseDetails())
+            .caseDetailsBefore(callbackRequest().getCaseDetailsBefore())
+            .build();
```
Comment (swalker125) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsControllerAboutToSubmitTest.java]: I would remove this

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsControllerAboutToSubmitTest.java [hunk 8]
```diff
@@ -56,28 +57,29 @@
...
 
-        CaseData caseData = mapper.convertValue(caseDetails.getCaseDetails().getData(), CaseData.class);
+        CaseData caseData = mapper.convertValue(callbackRequest.getCaseDetails().getData(), CaseData.class);
```
Comment (swalker125) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsControllerAboutToSubmitTest.java]:
```suggestion
        CaseData caseData = mapper.convertValue(callbackRequest().getCaseDetails().getData(), CaseData.class);
```

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsController.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsController.java]: can you confirm this?? I don't believe

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsController.java [hunk 9]
```diff
@@ -99,14 +99,24 @@ public AboutToStartOrSubmitCallbackResponse handleAboutToSubmitEvent(
...
 
         List<DocmosisTemplates> templateTypes = getProceedingTemplateTypes(caseData);
 
         List<Document> uploadedDocuments = generateAndUploadDocuments(userId, authorization, templateData,
             templateTypes);
 
-        caseDetails.getData().put("noticeOfProceedingsBundle", createNoticeOfProceedingsCaseData(uploadedDocuments));
+        List<Element> noticeOfProceedingCaseData = createNoticeOfProceedingsCaseData(uploadedDocuments);
+
+        if (callbackRequest.getCaseDetailsBefore().getData().get("noticeOfProceedingsBundle") != null) {
```
Comment (swalker125) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsController.java]: there's an isNotNull method that could be used here, not super important though

File: service/src/main/java/uk/gov/hmcts/reform/fpl/enums/DocmosisTemplates.java [hunk 10]
```diff
@@ -1,8 +1,8 @@
 package uk.gov.hmcts.reform.fpl.enums;
 
 public enum DocmosisTemplates {
-    C6("FL-PLW-HNO-ENG-00092.docx", "Notice_of_proceedings_(c6)"),
-    C6A("FL-PLW-HNO-ENG-00093.docx", "Notice_of_proceedings_(c6a)"),
+    C6("FL-PLW-HNO-ENG-00092.docx", "Notice_of_proceedings_c6"),
```
Comment (swalker125) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/enums/DocmosisTemplates.java]: is there a reason this isn't your pdf title?

Comment (swalker125) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/enums/DocmosisTemplates.java]: i.e c6.pdf or something...

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/NoticeOfProceedingsServiceTest.java [hunk 11]
```diff
@@ -31,8 +36,7 @@
 import static uk.gov.hmcts.reform.fpl.utils.CaseDataGeneratorHelper.createPopulatedChildren;
 
 @ExtendWith(SpringExtension.class)
-
-public class NoticeOfProceedingServiceTest {
+class NoticeOfProceedingsServiceTest {
     @SuppressWarnings({"membername", "AbbreviationAsWordInName"})
```
Comment (swalker125) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/NoticeOfProceedingsServiceTest.java]: any way we could remove these supressions?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/NoticeOfProceedingsServiceTest.java [hunk 12]
```diff
@@ -45,9 +49,33 @@
...
+    }
+
+    @Test
+    void shouldNotRetrieveExistingDocumentsAWhenTemplateListIncludeBothC6AndC6A() {
+        List<DocmosisTemplates> templatesList = ImmutableList.of(C6, C6A);
+        CaseData caseData = generateNoticeOfProceedingBundle(templatesList);
+        List<Element<DocumentBundle>> removedDocuments = noticeOfProceedingService
+            .getRemovedDocumentBundles(caseData, templatesList);
+
+        assertThat(removedDocuments).hasSize(0);
```
Comment (swalker125) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/NoticeOfProceedingsServiceTest.java]:
```suggestion
        assertThat(removedDocuments).isEmpty();
```

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsController.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (Jamesnelson117) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/NoticeOfProceedingsController.java]: Ah sorry I was thinking of the getRemovedDocumentBundle method, change will be made in next commit.

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/NoticeOfProceedingsServiceTest.java (hunk 11 above)
Comment (Jamesnelson117) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/NoticeOfProceedingsServiceTest.java]: I've a feeling this is no longer needed...

Comment (Jamesnelson117) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/NoticeOfProceedingsServiceTest.java]: Confirmed, removal has no impact.

**Comment by Jamesnelson117** [comment]
Do not merge until permissions to Notice of proceeding have been removed.

**Pull request merged**